lib/Dialect/TritonGPU/Transforms/Pipeliner/ScheduleLoops.cpp

```diff
@@ -205,22 +313,102 @@ CoarseSchedule scheduleKeyOps(scf::ForOp forOp,
   if (latOps.empty())
     return CoarseSchedule(0);
 
-  DominanceInfo domInfo(forOp);
+  // Schedule parallel dot pattern.
+  int maxStages = getNumStagesOrDefault(forOp, defaultNumStages);
   // Compute the longest path to the yield for each operation reachable
-  // from any latency operation.
+  // from any latency operation. We also use this to embed stage information
+  // for mmas.
   DenseMap<Operation *, int> distance;
-  std::function<int(Operation *)> computeDistance = [&](Operation *op) -> int {
+  // Track the MMA cluster information for the independent dot chain path.
+  DenseMap<Operation *, int> clusterMap;
+  auto [chains, success] = determineIndependentDotChains(forOp);
+  size_t maxChainLength = 0;
+  size_t numDots = 0;
+  SmallVector<int> maxClusterPerDistance(maxStages, -1);
+  if (success) {
+    for (auto &chain : chains) {
+      maxChainLength = std::max(maxChainLength, chain.size());
+      numDots += chain.size();
+    }
+    // Assign each chain in order. Any time we wrap around to the
+    // next stage we assign that op to a later stage. When we can
+    // get the same dot distance with a later stage (but an earlier cluster),
+    // then we will.
+    int startingOffset = 0;
+    int incrementValue = (numDots - (maxChainLength - 1));
+    for (auto &chain : chains) {
```

Comment:
See the PR description for the context on how this works. Happy to hop on a call tomorrow if needed to explain the reasoning.